mylib/src/lib.rs

```diff
@@ -0,0 +1,102 @@
+use {
+    anyhow::{anyhow, Result},
+    clap::Parser,
+    wasmtime::{Config, Engine, Linker, Module, Store},
+    wasmtime_wasi::{Dir, I32Exit, WasiCtx, WasiCtxBuilder},
+};
+
+fn parse_mapdir(s: &str) -> Result<(String, String)> {
+    if let Some((guest_dir, host_dir)) = s.split_once("::") {
+        Ok((guest_dir.into(), host_dir.into()))
+    } else {
+        Err(anyhow!(
+            "expected string of form GUEST_DIR::HOST_DIR; got {s}"
+        ))
+    }
+}
+
+fn parse_env(s: &str) -> Result<(String, String)> {
+    if let Some((name, value)) = s.split_once('=') {
+        Ok((name.into(), value.into()))
+    } else {
+        Err(anyhow!("expected string of form NAME=VALUE; got {s}"))
+    }
+}
+
+#[derive(Parser)]
+struct Options {
+    #[clap(long, value_name = "GUEST_DIR::HOST_DIR", value_parser = parse_mapdir)]
+    mapdir: Vec<(String, String)>,
+
+    #[clap(long, value_name = "NAME=VALUE", value_parser = parse_env)]
+    env: Vec<(String, String)>,
+}
+
+struct Ctx {
+    wasi: WasiCtx,
+}
+
+fn ignore_successful_proc_exit_trap(guest_err: anyhow::Error) -> Result<()> {
+    match guest_err.root_cause().downcast_ref::<I32Exit>() {
+        Some(trap) => match trap.0 {
+            0 => Ok(()),
+            _ => Err(guest_err),
+        },
+        None => Err(guest_err),
+    }
+}
+
+fn run_internal(guest: &[u8]) -> Result<()> {
+    let options = Options::parse();
```

Comment:
This probably won't do anything useful, since the Rust `std::env::args` won't have been initialized when using this as a library from C.  You could update `run_rust_wasm_native_binary` to accept CLI parameters and pass them to https://docs.rs/clap/latest/clap/trait.Parser.html#method.try_parse_from.  Or you could just delete all the options parsing and WASI init code if you don't care about giving the guest access to host environment variables and directories.

Reply:
Ah I see. Actually I do care about that, although I think we have not implemented that yet in our WASM backend. This poses an interesting problem --- in our LLVM backend we allow access to the whole file system. Here we have a nice option to actually restrict it, it's actually not clear to me right now what the default should be.